File: src/thietkemoi/thanxe_new_description/thanxe_new_description/depth_heatmap.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np


class DepthHeatmap(Node):

    def __init__(self):
        super().__init__('depth_heatmap')

        self.bridge = CvBridge()

        self.sub = self.create_subscription(
            Image,
            '/base_camera/base_camera_sensor/depth/image_raw',
            self.callback,
            10
        )

        self.pub = self.create_publisher(
            Image,
            '/base_camera/depth_heatmap',
            10
        )

        self.get_logger().info('Depth heatmap node started.')

    def callback(self, msg):
        depth = self.bridge.imgmsg_to_cv2(msg, desired_encoding='passthrough')

        # Convert to float32 if needed
        depth = np.float32(depth)

        # Replace inf/nan with 0
        depth = np.where(np.isfinite(depth), depth, 0.0)

        # Normalize to 0-255
        d_min = np.min(depth[depth > 0]) if np.any(depth > 0) else 0.0
        d_max = np.max(depth)

        if d_max - d_min > 0:
            normalized = (depth - d_min) / (d_max - d_min)
        else:
            normalized = np.zeros_like(depth)

        # Convert to uint8 (0 = near, 255 = far)
        gray = (normalized * 255).astype(np.uint8)

        # Apply colormap: COLORMAP_JET maps 0->blue, 255->red
        # We flip so that near=red, far=blue
        gray_flipped = 255 - gray
        heatmap = cv2.applyColorMap(gray_flipped, cv2.COLORMAP_JET)

        # Set invalid depth (0) to black
        heatmap[depth <= 0] = [0, 0, 0]

        # Publish
        heatmap_msg = self.bridge.cv2_to_imgmsg(heatmap, encoding='bgr8')
        heatmap_msg.header = msg.header
        self.pub.publish(heatmap_msg)


def main(args=None):
    rclpy.init(args=args)
    node = DepthHeatmap()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/my_ur_control_cpp/src/depth_heatmap.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

class DepthHeatmap : public rclcpp::Node {
public:
    DepthHeatmap() : Node("depth_heatmap") {
        sub_ = create_subscription<sensor_msgs::msg::Image>(
            "/base_camera/base_camera_sensor/depth/image_raw", 10,
            std::bind(&DepthHeatmap::callback, this, std::placeholders::_1));
        pub_ = create_publisher<sensor_msgs::msg::Image>("/base_camera/depth_heatmap", 10);
        RCLCPP_INFO(this->get_logger(), "Depth heatmap C++ node started.");
    }

private:
    void callback(const sensor_msgs::msg::Image::SharedPtr msg) {
        cv::Mat depth;
        try {
            // Ép kiểu về Float32 để tính toán y hệt numpy
            depth = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::TYPE_32FC1)->image;
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        // Thay thế NaN và Inf bằng 0.0
        cv::patchNaNs(depth, 0.0);
        
        // Tạo mask cho các giá trị > 0
        cv::Mat valid_mask = depth > 0;
        
        double d_min, d_max;
        cv::minMaxLoc(depth, &d_min, &d_max, nullptr, nullptr, valid_mask);

        cv::Mat normalized = cv::Mat::zeros(depth.size(), CV_32FC1);
        if (d_max - d_min > 0) {
            normalized = (depth - d_min) / (d_max - d_min);
        }

        // Chuyển sang 8-bit (0-255)
        cv::Mat gray;
        normalized.convertTo(gray, CV_8UC1, 255.0);

        // Lật màu: gần = đỏ, xa = xanh (255 - gray)
        cv::Mat heatmap;
        cv::applyColorMap(gray, heatmap, cv::COLORMAP_JET);

        // Set các điểm ảnh không hợp lệ (depth <= 0) thành màu đen
        heatmap.setTo(cv::Scalar(0, 0, 0), depth <= 0);

        auto out_msg = cv_bridge::CvImage(msg->header, "bgr8", heatmap).toImageMsg();
        pub_->publish(*out_msg);
    }
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DepthHeatmap>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}